src/paymasters/SignaturePaymaster.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import "@aa/core/BasePaymaster.sol";
+import "@aa/core/UserOperationLib.sol";
+import "@aa/core/UserOperationLib.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+/**
+ * @notice
+ * A signature paymaster that uses external service to decide whether to pay for the UserOp.
+ * The paymaster trusts an external signer to sign the transaction.
+ * The calling user must pass the UserOp to that external signer first, which performs
+ * whatever off-chain verification before signing the UserOp.
+ * Note that this signature is NOT a replacement for the account-specific signature:
+ * - the paymaster checks a signature to agree to PAY for GAS.
+ * - the account checks a signature to prove identity and account ownership.
+ */
+contract SignaturePaymaster is BasePaymaster, Initializable, UUPSUpgradeable {
+    using UserOperationLib for UserOperation;
+    using ECDSA for bytes32;
+
+    address public immutable verifyingSigner;
+
+    uint256 private constant VALID_TIMESTAMP_OFFSET = 20;
+    uint256 private constant SIGNATURE_OFFSET = 84;
+
+    mapping(address => uint256) public senderNonce;
+
+    // ========== Constructor & Upgrades ============
+
+    constructor(IEntryPoint _entryPoint, address _verifyingSigner) BasePaymaster(_entryPoint) {
+        _disableInitializers();
+        verifyingSigner = _verifyingSigner;
```

Comment:
CMIIW, i think this should be on the `initialize` on we should avoid `immutable` variables when using UUPSUpgradeable

Reply:
We had a nice discussion about this in other channel.

Reply:
yep, thanks for the clarifications!